ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py

```diff
@@ -190,25 +149,129 @@ def remove(self, app, leaveObjects=False):
 
                 self.LOG.info('Removing {} relationships from existing devices.'.format(self.id))
-                self._buildDeviceRelations()
+                self._buildDeviceRelations(app)
 
             for dcname, dcspec in self.device_classes.iteritems():
                 if dcspec.remove:
                     self.remove_device_class(app, dcspec)
 
             # Remove EventClasses with remove flag set
-            self.remove_organizer_or_subs(app.dmd.Events,
+            self.remove_organizer_or_subs(app.zport.dmd.Events,
                                           self.event_classes,
                                           'mappings',
                                           'removeInstances')
             # Remove Process Classes/Organizers with remove flag set
-            self.remove_organizer_or_subs(app.dmd.Processes,
+            self.remove_organizer_or_subs(app.zport.dmd.Processes,
                                           self.process_class_organizers,
                                           'process_classes',
                                           'removeOSProcessClasses')
 
         super(ZenPack, self).remove(app, leaveObjects=leaveObjects)
 
+    def update_object(self, app, parent, relname, object_id, spec, specparam):
+        """Compare object to be installed to existing objects, optionally creating the new object"""
+        # this backup should exist if previous installed version of this zenpack uses ZPL 2.0
+        backup_target = self.get_object(parent, relname, "{}-backup".format(object_id))
+        # otherwise this is the existing object pre-zpl 2.0
+        existing_target = self.get_object(parent, relname, object_id)
+        # we'll use whichever we get
+        candidate_target = backup_target or existing_target
+
+        if not candidate_target:
+            # if no candidate target is found, then this is new to this zenpack
+            spec.create(app.zport.dmd)
+        else:
+            # check the difference between our candidate and the new spec
+            # and back up the candidate if there is a difference
+            diff = self.check_diff(app, parent, relname, candidate_target, spec, specparam)
+            # if there was a difference, keep the backup and create the new template
+            if diff:
+                spec.create(app.zport.dmd)
+            else:
+                # otherwise just return the backup template to its original location
+                if backup_target:
+                    self.move_object(parent, relname, backup_target.id, object_id)
+                # or in the case of the existing object, leave it alone
+                else:
+                    pass
+
+    def check_diff(self, app, parent, relname, object, spec, specparam):
+        """Return True if object has changed creating preupgrade backup if needed"""
+        diff = self.object_changed(app, object, spec, specparam)
+        if diff:
```

Comment:
You could rearrange this a bit to eliminate a level of indentation for the method.

```python
if not diff:
    return False

# preserve the existing object if different
# ..etc.
```